File: src/advent_2016/Day10.java
package advent_2016;

import advent_2016.day_10.*;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day10 {
    private static Map<String, Bot> bots = new HashMap<>();
    private static Map<String, Bin> bins = new HashMap<>();

    private static void step1() {
        bots.values().stream()
            .filter(bot -> bot.getComparisons().stream()
                .anyMatch(comparison -> comparison.getLow() == 17 && comparison.getHigh() == 61))
            .map(Bot::toString)
            .forEach(System.out::println);
    }

    private static void step2() {
        System.out.println(bins.get("0").getValues().get(0) *
            bins.get("1").getValues().get(0) *
            bins.get("2").getValues().get(0));
    }

    public static void main(String[] args) {
        Utils.readFile("day10.txt")
            .sorted()
            .forEach(Day10::matchInstruction);
        step1();
        step2();
    }

    private static void matchInstruction(String instruction) {

        Matcher split = Pattern.compile("bot (\\d+) gives low to ([a-z]+) (\\d+) and high to ([a-z]+) (\\d+)", Pattern.CASE_INSENSITIVE)
            .matcher(instruction);

        if (split.matches()) {
            String number = split.group(1);
            Bot bot = getOrPut(bots, number, new Bot(number));
            bot.instruct(getReceiver(split.group(2), split.group(3)), getReceiver(split.group(4), split.group(5)));

            return;
        }

        Matcher value = Pattern.compile("value (\\d+) goes to ([a-z]+) (\\d+)", Pattern.CASE_INSENSITIVE)
            .matcher(instruction);

        if (value.matches()) {
            Receiver receiver = getReceiver(value.group(2), value.group(3));
            receiver.receive(Integer.parseInt(value.group(1)));

            return;
        }

        throw new UnknownInstructionException(instruction);
    }

    private static Receiver getReceiver(String type, String number) {
        switch (type) {
            case "bot":
                return getOrPut(bots, number, new Bot(number));
            case "output":
                return getOrPut(bins, number, new Bin(number));
            default:
                return null;
        }
    }

    private static <T> T getOrPut(Map<String, T> map, String key, T value) {
        T existing = map.get(key);
        if (existing == null) {
            map.put(key, value);
            return value;
        }
        return existing;
    }
}

File: src/advent_2016/day_10.clj
(ns advent-2016.day-10
    (:require [advent-2016.utils.core :as u]))

(def ^:private bots (atom {}))

(def ^:private bins (atom {}))

(defn ^:private get-receiver [type number]
    (-> (if (= type "bot") bots bins)
        (deref)
        (get number)
        (:receive identity)))

(defn ^:private make-bot [name low-to-type low-to-num high-to-type high-to-num]
    (let [processed    (atom [])
          stored-value (atom nil)]
        {:receive   (fn [value]
                        (if @stored-value
                            (let [[min' max'] (map #(% value @stored-value) [min max])]
                                ((get-receiver low-to-type low-to-num) min')
                                ((get-receiver high-to-type high-to-num) max')
                                (reset! stored-value nil)
                                (swap! processed conj [min' max']))
                            (reset! stored-value value)))
         :name      name
         :processed (fn [] @processed)}))

(defn ^:private make-bin [name]
    (let [values (atom [])]
        {:receive   (fn [value]
                        (swap! values conj value))
         :name      name
         :processed (fn [] @values)}))

(defn ^:private build [line]
    (let [[_ bot-number low-to-type low-to-number high-to-type high-to-number]
          (re-find #"bot (\d+) gives low to ([a-z]+) (\d+) and high to ([a-z]+) (\d+)" line)]
        (cond
            (= "output" low-to-type) (swap! bins assoc low-to-number (make-bin low-to-number))
            (= "output" high-to-type) (swap! bins assoc high-to-number (make-bin high-to-number)))
        (swap! bots assoc bot-number (make-bot bot-number low-to-type low-to-number high-to-type high-to-number))))

(defn ^:private run [line]
    (let [[_ value type number]
          (re-find #"value (\d+) goes to ([a-z]+) (\d+)" line)]
        ((get-receiver type number) (u/parse-int value))))

(defn ^:private find-processor [bots low high]
    (->> bots
        (vals)
        (filter (comp #(contains? % [low high]) set #(%) :processed))
        (first)
        (:name)))

(defn ^:private build-and-run [instructions]
    (doall (map build (get instructions \b)))
    (doall (map run (get instructions \v))))

;; 116
(defn step-1 []
    (->> (u/read-file 10 #"\n")
        (group-by first)
        (build-and-run))
    (find-processor @bots 17 61))
